New CreateEdit documentation (#7424)
marmelab/react-admin: docs/Edit.md

---
layout: default
title: "The Edit Component"
---

# `<Edit>`

> The `<Edit>` component is the main component for edition pages. It fetches a record based on the URL, preparess a form submit handler, and renders the page title and actions. It is not responsible for rendering the actual form - that's the job of its child component (usually a form component, like [`<SimpleForm>`](./SimpleForm.md)). This form component uses its children ([`<Input>`](./Inputs.md) components) to render each form input.

Reviewer: ```suggestion
The `<Edit>` component is the main component for edition pages. It fetches a record based on the URL, prepares a form submit handler, and renders the page title and actions. It is not responsible for rendering the actual form - that's the job of its child component (usually a form component, like [`<SimpleForm>`](./SimpleForm.md)). This form component uses its children ([`<Input>`](./Inputs.md) components) to render each form input.
```

            <TextInput source="title" validate={required()} />
            <TextInput multiline source="teaser" validate={required()} />
            <RichTextInput source="body" validate={required()} />
            <DateInput label="Publication date" source="published_at" />
            <ReferenceManyField label="Comments" reference="comments" target="post_id">
                <Datagrid>
                    <TextField source="body" />
                    <DateField source="created_at" />
                    <EditButton />
                </Datagrid>
            </ReferenceManyField>
        </SimpleForm>
    </Edit>
);

// in src/App.js
import * as React from "react";
import { Admin, Resource } from 'react-admin';
import jsonServerProvider from 'ra-data-json-server';

import { PostEdit } from './posts';

const App = () => (
    <Admin dataProvider={jsonServerProvider('https://jsonplaceholder.typicode.com')}>
        <Resource name="posts" edit={PostEdit} />
    </Admin>
);

export default App;
```

You can customize the `<Edit>` component using the following props:

* [`actions`](#actions): override the actions toolbar with a custom component

> * [`aside`](#aside-component): component to render aside to the main content

Reviewer: ```suggestion
* [`aside`](#aside): component to render aside to the main content
```

* `children`: the components that renders the form
* `className`: passed to the root component
* [`component`](#component): override the root component

> * [`disableAuthentication`](#disable-authentication): disable the authentication check

Reviewer: ```suggestion
* [`disableAuthentication`](#disableauthentication): disable the authentication check
```

* [`id`](#id): the id of the record to edit

> * [`mutationMode`](#mutation-mode): switch to optimistic or pessimistic mutations (undoable by default)

Reviewer: ```suggestion
* [`mutationMode`](#mutationmode): switch to optimistic or pessimistic mutations (undoable by default)
```

Some form layouts also use `Card`, in which case the user ends up seeing a card inside a card, which is bad UI. To avoid that, you can override the main page container by passing a `component` prop :

```jsx
// use a div as root component
const PostEdit = () => (
    <Edit component="div">
        ...
    </Edit>
);

// use a custom component as root component 
const PostEdit = () => (
    <Edit component={MyComponent}>
        ...
    </Edit>
);
```

The default value for the `component` prop is `Card`.

## `disableAuthentication`

By default, the `<Edit>` component will automatically redirect the user to the login page if the user is not authenticated. If you want to disable this behavior and allow anonymous access to a creation page, set the `disableAuthentication` prop to `true`.

```jsx
const PostEdit = () => (
    <Edit disableAuthentication>
        ...
    </Edit>
);
```

## `id`

Components based on `<Edit>` are often used as `<Resource edit>` props, and therefore rendered when the URL matches `/[resource]/[id]`. The `<Edit>` component generates a call to `dataProvider.update()` using the id from the URL by default.

> You can decide to use a `<Edit>` component in another path, or embedded in a page editing a related record (e.g. in a Dialog). In that canse, you can explicitely set the `id` value:

Reviewer: ```suggestion
You can decide to use a `<Edit>` component in another path, or embedded in a page editing a related record (e.g. in a Dialog). In that case, you can explicitly set the `id` value:
```

```jsx
import * as React from 'react';
import { useNotify, useRefresh, useRedirect, Edit, SimpleForm } from 'react-admin';

const PostEdit = () => {
    const notify = useNotify();
    const refresh = useRefresh();
    const redirect = useRedirect();

  const onSuccess = (data) => {
        notify(`Changes to post "${data.title}" saved`);
        redirect('/posts');
        refresh();
    };

    return (
        <Edit mutationOptions={{ onSuccess }} mutationMode="pessimistic">
            <SimpleForm>
                ...
            </SimpleForm>
        </Edit>
    );
}
```

**Tip**: If you want to have different success side effects based on the button clicked by the user (e.g. if the creation form displays two submit buttons, one to "save and redirect to the list", and another to "save and display an empty form"), you can set the `mutationOptions` prop on the `<SaveButton>` component, too.

> Similarly, you can override the failure side effects with an `onError` option. By default, when the save action fails at the dataProvider level, react-admin shows an error notification.

Reviewer: ```suggestion
Similarly, you can override the failure side effects with an `onError` option. By default, when the save action fails at the dataProvider level, react-admin shows a notification error.
```

## `record`

On mount, the `<Edit>` component will fetch the record to edit using the `resource` and the `id` from the URL. You can also use `<Edit>` independently of the URL, by passing a record directly via the `record` prop.

{% raw %}
```jsx
const PostEdit = () => (
    <Edit record={{ id: 123, title: 'Hello' }}>
        <SimpleForm>
            ...
        </SimpleForm>
    </Edit>
);
```
{% endraw %}

This is useful when you want to embed an edition view in another view, for instance when displaying the edit form in a dialog. 

## `redirect`

By default, submitting the form in the `<Edit>` view redirects to the `<List>` view.

You can customize the redirection by setting the `redirect` prop to one of the following values:

- `'list'`: redirect to the List view (the default)
- `'show'`: redirect to the Show view
- `false`: do not redirect

```jsx
const PostEdit = () => (
    <Edit redirect="show">
        ...
    </Edit>
);
```

Note that the `redirect` prop is ignored if you set [the `mutationOptions` prop](#mutationoptions). See that prop for how to set a different redirection path in that case. 

## `resource`

> Components based on `<Edit>` are often used as `<Resource edit>` props, and therefore rendered when the URL matches `/[resource]/[id]`. The `<Edit>` component generates a call to `dataProvider.update()` using the ressource name from the URL by default.

Reviewer: ```suggestion
Components based on `<Edit>` are often used as `<Resource edit>` props, and therefore rendered when the URL matches `/[resource]/[id]`. The `<Edit>` component generates a call to `dataProvider.update()` using the resource name from the URL by default.
```

> You can decide to use a `<Edit>` component in another path, or embedded in a page using another ressource name (e.g. in a Dialog). In that canse, you can explicitely set the `ressource` name:

Reviewer: ```suggestion
You can decide to use a `<Edit>` component in another path, or embedded in a page using another resource name (e.g. in a Dialog). In that case, you can explicitly set the `resource` name:
```

**Tip**: If you want a confirmation dialog for the Delete button but don't mind undoable Edits, then pass a [custom toolbar](./SimpleForm.md#toolbar) to the form, as follows:

{% raw %}
```jsx
import * as React from "react";
import {
    Toolbar,
    SaveButton,
    DeleteButton,
    Edit,
    SimpleForm,
} from 'react-admin';

const CustomToolbar = props => (
    <Toolbar
        {...props}
        sx={{ display: 'flex', justifyContent: 'space-between' }}
    >
        <SaveButton />
        <DeleteButton mutationMode="pessimistic" />
    </Toolbar>
);

const PostEdit = () => (
    <Edit>
        <SimpleForm toolbar={<CustomToolbar />}>
            ...
        </SimpleForm>
    </Edit>
);
```
{% endraw %}

## `mutationOptions`

`<Edit>` calls `dataProvider.update()` via react-query's `useMutation` hook. You can customize the options you pass to this hook, e.g. to override success or error side effects, by setting the `mutationOptions` prop. Refer to the [useMutation documentation](https://react-query.tanstack.com/reference/useMutation) in the react-query website for a list of the possible options.

> Let's see an example with the success side effect. By default, when the save action succeeds, react-admin shows a notification, and redirects to another page. You can override this behavior and pass custom success side effects by providing a `mutationOptions` prop with an `onSuccess` key:

Reviewer 2: ```suggestion
Let's see an example with the success side effect. By default, when the save action succeeds, react-admin shows a notification, and redirects to the list page. You can override this behavior and pass custom success side effects by providing a `mutationOptions` prop with an `onSuccess` key:
```

For instance, you can force a refetch on reconnect:

{% raw %}
```jsx
const PostEdit = () => (
    <Edit queryOptions={{ refetchOnReconnect: true }}>
        ...
    </Edit>
);
```
{% endraw %}

Refer to the [useQuery documentation](https://react-query.tanstack.com/reference/useQuery) in the react-query website for a list of the possible options.

## `record`

On mount, the `<Edit>` component will fetch the record to edit using the `resource` and the `id` from the URL. You can also use `<Edit>` independently of the URL, by passing a record directly via the `record` prop.

{% raw %}
```jsx
const PostEdit = () => (
    <Edit record={{ id: 123, title: 'Hello' }}>
        <SimpleForm>
            ...
        </SimpleForm>
    </Edit>
);
```
{% endraw %}

This is useful when you want to embed an edition view in another view, for instance when displaying the edit form in a dialog. 

## `redirect`

By default, submitting the form in the `<Edit>` view redirects to the `<List>` view.

You can customize the redirection by setting the `redirect` prop to one of the following values:

- `'list'`: redirect to the List view (the default)
- `'show'`: redirect to the Show view

> - `false`: do not redirect

Reviewer 2: We also accept a function `redirect((resource, id, data) => string | Path)`
